src/Stratis.Bitcoin/Consensus/PeerGetHeaderRequestBehavior.cs

```diff
@@ -0,0 +1,132 @@
+            this.peerBanning = peerBanning;
+        }
+
+        public override object Clone()
+        {
+            return new RateLimitingBehavior(this.dateTimeProvider, this.loggerFactory, this.peerBanning);
+        }
+
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceived, true);
+        }
+
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceived);
+        }
+
+        /// <summary>
+        /// Processes an incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        private Task OnMessageReceived(INetworkPeer peer, IncomingMessage message)
+        {
+            switch (message.Message.Payload)
+            {
+                case GetHeadersPayload getHeaders:
+                    this.AssessPeerGetHeaderFrequency(getHeaders);
+                    break;
+
+                default:
+                    break;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void AssessPeerGetHeaderFrequency(GetHeadersPayload getHeaders)
+        {
+            // Is the last requested hash the same as this request.
+            if (this.getHeaderLastRequestHash == getHeaders.BlockLocator.Blocks.First())
+            {
+                this.logger.LogDebug($"{this.AttachedPeer.PeerEndPoint} block locator matches previous, count {this.getHeaderRequestCount}");
+
+                // Was this hash requested less than 60 seconds ago.
+                if (this.getHeaderLastRequestedTimestamp > this.dateTimeProvider.GetUtcNow().AddSeconds(-60))
+                {
+                    this.getHeaderRequestCount++;
+                }
+                else
+                {
+                    this.getHeaderLastRequestedTimestamp = this.dateTimeProvider.GetUtcNow();
+                    this.getHeaderRequestCount = 0;
+                    return;
+                }
+
+                // If the same header was requested more than 3 times in the last 60 seconds,
+                // ban and disconnect the peer for 1 hour.
+                if (this.getHeaderRequestCount > GetHeaderRequestCountThreshold)
+                {
+                    this.peerBanning.BanAndDisconnectPeer(this.AttachedPeer.PeerEndPoint, BanDurationSeconds, $"Banned via {this.GetType().Name} for {BanDurationSeconds} seconds.");
+                    this.logger.LogDebug($"{this.AttachedPeer.PeerEndPoint} banned via {this.GetType().Name} for {BanDurationSeconds} seconds.");
+                }
+            }
+            else
+            {
+                this.getHeaderLastRequestHash = getHeaders.BlockLocator.Blocks.First();
+                this.getHeaderRequestCount = 0;
```

Comment:
should be set to 1

Reply:
OK so here we reset the count to 0 as more than 60s has passed since he last asked for the same set of headers.

Reply:
it is a counter that shows how many getheaders messages we've received with first hash equal to `getHeaderLastRequestHash`

Right now we will ban after we receive 12 same getheaders. 

If you wanted to ban after 10 same messages then this should be 1 and on 119 it should be >=

Reply:
Actually, the fix here is that `if (this.getHeaderRequestCount > GetHeaderRequestCountThreshold)`
should be `if (this.getHeaderRequestCount >= GetHeaderRequestCountThreshold)`.

```
Before After ; Condition Check
0 1; 1 >= 10;
1 2; 2 >= 10;
2 3; 3 >= 10;
3 4; 4 >= 10;
4 5; 5 >= 10;
5 6; 6 >= 10;
6 7; 7 >= 10;
7 8; 8 >= 10;
8 9; 9 >= 10;
9 10; 10 >= 10;
```

Reply:
Ah well spotted @noescape00